src/cocotb/clock.py

```
@@ -109,52 +115,60 @@ async def custom_clock():
     """
 
     def __init__(
-        self, signal, period: Union[float, Fraction, Decimal], units: str = "step"
+        self,
+        signal,
+        period: Union[float, Fraction, Decimal],
+        units: str = "step",
+        impl: str = "auto",
     ):
         self.signal = signal
         self.period = get_sim_steps(period, units)
-        self.half_period = get_sim_steps(period / 2, units)
         self.frequency = 1 / get_time_from_sim_steps(self.period, units="us")
-        self.hdl = None
-        self.signal = signal
-        self.coro = None
-        self.mcoro = None
+        self.t_high = self.period // 2
+
+        # If not using GpiClock, clkobj is None
+        if impl == "auto":
+            impl = "gpi" if cocotb._conf.trust_inertial else "py"
+        if impl == "gpi":
+            self.clkobj = clock_create(self.signal._handle)
+        elif impl == "py":
+            self.clkobj = None
+        else:
+            raise ValueError(f"Invalid clock impl '{impl}'")
 
-    async def start(
-        self, cycles: Optional[int] = None, start_high: bool = True
-    ) -> None:
+    async def start(self, start_high: bool = True) -> None:
         r"""Clocking coroutine.  Start driving your clock by :func:`cocotb.start`\ ing a
         call to this.
 
         Args:
-            cycles: Cycle the clock *cycles* number of times,
-                or if ``None`` then cycle the clock forever.
-                Note: ``0`` is not the same as ``None``, as ``0`` will cycle no times.
             start_high: Whether to start the clock with a ``1``
                 for the first half of the period.
                 Default is ``True``.
 
                 .. versionadded:: 1.3
         """
-        t = Timer(self.half_period)
-        if cycles is None:
-            it = itertools.count()
+        if self.clkobj:
+            phase = 0 if start_high else self.t_high
+            self.clkobj.start(self.period, self.t_high, phase)
```

Check the output of the call to `start` here.

---

Can you add a bit more detail here? Specifically, I think a failure should always throw (added a test), and it seems that we would just let the exception propagate.

---

No this is good. Other functions in this module return `None` rather than throw an exception, so I was making an incorrect assumption here. Looks good the way it is.